File: packages/sdk-java/src/main/java/xyz/alphahuman/sdk/AlphahumanMemoryClient.java
package xyz.alphahuman.sdk;

import xyz.alphahuman.sdk.internal.Json;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;

/**
 * Client for the Alphahuman Memory API.
 * <p>
 * Requires Java 11+. Zero runtime dependencies.
 */
public class AlphahumanMemoryClient implements AutoCloseable {

    private static final String DEFAULT_BASE_URL = "https://staging-api.alphahuman.xyz";
    private static final String BASE_URL_ENV = "ALPHAHUMAN_BASE_URL";

    private final String baseUrl;
    private final String token;
    private final HttpClient httpClient;

    public AlphahumanMemoryClient(String token) {
        this(token, null);
    }

    public AlphahumanMemoryClient(String token, String baseUrl) {
        if (token == null || token.trim().isEmpty()) {
            throw new IllegalArgumentException("token is required");
        }
        this.token = token;

        String resolved = baseUrl;
        if (resolved == null || resolved.isEmpty()) {
            resolved = System.getenv(BASE_URL_ENV);
        }
        if (resolved == null || resolved.isEmpty()) {
            resolved = DEFAULT_BASE_URL;
        }
        this.baseUrl = resolved.replaceAll("/+$", "");

        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(30))
                .build();
    }

    /** Insert (ingest) a document into memory. POST /v1/memory/insert */
    public InsertMemoryResponse insertMemory(InsertMemoryParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/v1/memory/insert", body);
        return InsertMemoryResponse.fromMap(response);
    }

    /** Recall context from Master node. POST /v1/memory/recall */
    public RecallMemoryResponse recallMemory(RecallMemoryParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/v1/memory/recall", body);
        return RecallMemoryResponse.fromMap(response);
    }

    /** Delete memory (admin). POST /v1/memory/admin/delete */
    public DeleteMemoryResponse deleteMemory(DeleteMemoryParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/v1/memory/admin/delete", body);
        return DeleteMemoryResponse.fromMap(response);
    }

    /** Query memory via RAG. POST /v1/memory/query */
    public QueryMemoryResponse queryMemory(QueryMemoryParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/v1/memory/query", body);
        return QueryMemoryResponse.fromMap(response);
    }

    /** Recall memories from Ebbinghaus bank. POST /v1/memory/memories/recall */
    public RecallMemoriesResponse recallMemories(RecallMemoriesParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/v1/memory/memories/recall", body);
        return RecallMemoriesResponse.fromMap(response);
    }

    @Override
    public void close() {
        // HttpClient does not require explicit close in Java 11
    }

    // ---- Internal ----

    private Map<String, Object> post(String path, Map<String, Object> body) {
        String jsonBody = Json.serialize(body);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + path))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + token)
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                .timeout(Duration.ofSeconds(30))
                .build();

        HttpResponse<String> response;
        try {
            response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            throw new RuntimeException("HTTP request failed: " + e.getMessage(), e);
        }

        return handleResponse(response);
    }

    private Map<String, Object> handleResponse(HttpResponse<String> response) {
        String text = response.body();
        Map<String, Object> json;
        try {
            json = (text != null && !text.isEmpty()) ? Json.parse(text) : Map.of();
        } catch (Exception e) {
            throw new AlphahumanError(
                    "HTTP " + response.statusCode() + ": non-JSON response",
                    response.statusCode(),
                    text
            );
        }

        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            String message = "HTTP " + response.statusCode();
            Object error = json.get("error");
            if (error instanceof String) {
                message = (String) error;
            }
            throw new AlphahumanError(message, response.statusCode(), json);
        }

        return json;
    }
}

File: packages/sdk-csharp/src/Alphahuman.Sdk/AlphahumanMemoryClient.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Alphahuman.Sdk;

public sealed class AlphahumanMemoryClient : IDisposable
{
    private const string DefaultBaseUrl = "https://staging-api.alphahuman.xyz";

    private readonly string _token;
    private readonly string _baseUrl;
    private readonly HttpClient _httpClient;
    private bool _disposed;

    public AlphahumanMemoryClient(string token, string? baseUrl = null)
    {
        if (string.IsNullOrWhiteSpace(token))
            throw new ArgumentException("token is required");

        _token = token;
        _baseUrl = (baseUrl
            ?? Environment.GetEnvironmentVariable("ALPHAHUMAN_BASE_URL")
            ?? DefaultBaseUrl).TrimEnd('/');

        _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
    }

    internal AlphahumanMemoryClient(string token, string baseUrl, HttpClient httpClient)
    {
        if (string.IsNullOrWhiteSpace(token))
            throw new ArgumentException("token is required");

        _token = token;
        _baseUrl = baseUrl.TrimEnd('/');
        _httpClient = httpClient;
    }

    public async Task<InsertMemoryResponse> InsertMemoryAsync(InsertMemoryParams p)
    {
        p.Validate();
        var result = await PostAsync("/v1/memory/insert", p.ToJsonObject());
        return InsertMemoryResponse.FromJson(result);
    }

    public async Task<RecallMemoryResponse> RecallMemoryAsync(RecallMemoryParams? p = null)
    {
        p ??= new RecallMemoryParams();
        p.Validate();
        var result = await PostAsync("/v1/memory/recall", p.ToJsonObject());
        return RecallMemoryResponse.FromJson(result);
    }

    public async Task<DeleteMemoryResponse> DeleteMemoryAsync(DeleteMemoryParams? p = null)
    {
        p ??= new DeleteMemoryParams();
        p.Validate();
        var result = await PostAsync("/v1/memory/admin/delete", p.ToJsonObject());
        return DeleteMemoryResponse.FromJson(result);
    }

    public async Task<QueryMemoryResponse> QueryMemoryAsync(QueryMemoryParams p)
    {
        p.Validate();
        var result = await PostAsync("/v1/memory/query", p.ToJsonObject());
        return QueryMemoryResponse.FromJson(result);
    }

    public async Task<RecallMemoriesResponse> RecallMemoriesAsync(RecallMemoriesParams? p = null)
    {
        p ??= new RecallMemoriesParams();
        p.Validate();
        var result = await PostAsync("/v1/memory/memories/recall", p.ToJsonObject());
        return RecallMemoriesResponse.FromJson(result);
    }

    private async Task<JsonElement> PostAsync(string path, Dictionary<string, object?> body)
    {
        var json = JsonSerializer.Serialize(body, new JsonSerializerOptions
        {
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        });

        var request = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}{path}")
        {
            Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json"),
        };
        request.Headers.Authorization =
            new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _token);

        var response = await _httpClient.SendAsync(request);
        var responseBody = await response.Content.ReadAsStringAsync();
        return HandleResponse((int)response.StatusCode, responseBody);
    }

    private static JsonElement HandleResponse(int statusCode, string body)
    {
        JsonElement root;
        try
        {
            using var doc = JsonDocument.Parse(body);
            root = doc.RootElement.Clone();
        }
        catch (JsonException)
        {
            throw new AlphahumanError($"HTTP {statusCode}: non-JSON response", statusCode, body);
        }

        if (statusCode < 200 || statusCode >= 300)
        {
            var message = root.TryGetProperty("error", out var err)
                ? err.GetString() ?? $"HTTP {statusCode}"
                : $"HTTP {statusCode}";
            throw new AlphahumanError(message, statusCode, body);
        }

        return root;
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _httpClient.Dispose();
            _disposed = true;
        }
    }
}
